design/ceph-volume-provisioning.md

```diff
@@ -0,0 +1,122 @@
+# ceph-volume OSD Provisioning
+
+**Targeted for v0.9**
+
+Provisioning OSDs today is done directly by Rook. This needs to be simplified and improved by building
+on the functionality provided by the `ceph-volume` tool that is included in the ceph image.
+
+## Legacy Design
+As Rook is implemented today, the provisioning has a lot of complexity around:
+- Partitioning of devices for bluestore
+- Partitioning and configuration of a `metadata` device where the WAL and DB are placed on a different device from the data
+- Support for both directories and devices
+- Support for bluestore and filestore
+
+Since this is mostly handled by `ceph-volume` now, Rook should replace its own provisioning code and rely on `ceph-volume`.
+
+## ceph-volume Design
+
+`ceph-volume` is a CLI tool included in the `ceph/ceph` image that will be used to configure and run Ceph OSDs.
+`ceph-volume` will replace the OSD provisioning mentioned previously in the legacy design.
+
+At a high level this flow remains unchanged from the flow in the [one-osd-per-pod design](dedicated-osd-pod.md#create-new-osds).
+No new jobs or pods need to be launched from what we have today. The sequence of events in the OSD provisioning will be the following.
+- The cluster CRD specifies what nodes/devices to configure with OSDs
+- The operator starts a provisioning job on each node where OSDs are to be configured
+- The provisioning job:
+  - Detects what devices should be configured
+  - Calls `ceph-volume lvm batch` to prepare the OSDs on the node. A single call is made with all of the devices unless more specific settings are included for LVM and partitions.
+  - Calls `ceph-volume lvm list` to retrieve the results of the OSD configuration. Store the results in a configmap for the operator to take the next step.
+- The operator starts a deployment for each OSD that was provisioned. `rook` is the entrypoint for the container.
+  - `ceph-volume` is called to `activate` the osd, which mounts the config directory such as `/var/lib/ceph/osd-0`, using a tempfs mount
```

Comment:
Activate needs a bunch of info to work properly, like:

* the object store (`--bluestore` or `--filestore`)
* an  `OSD_ID`
* an `OSD_FSID`

So we must run `lvm list` before doing the activation, see how we do it for ceph-container here: https://github.com/ceph/ceph-container/blob/b8a8e159ffc4136099150bc9a7d7058e4cb158d5/src/daemon/osd_scenarios/osd_volume_activate.sh#L8-L25

Reply:
ok, i'll clarify this step. The `lvm list` step was done in the provisioning job a couple steps above. This will retrieve all the necessary info and put it in a configmap, which will be retrieved to pass to the activate.

Reply:
I've added it now. I was leaning towards leaving it out of the design doc to keep it higher level, but it's a simple enough detail to include.